In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

Dataset: https://archive.ics.uci.edu/ml/machine-learning-databases/undocumented/connectionist-bench/sonar/

In [2]:
data=pd.read_csv("Sonar.csv",header=None)

In [3]:
data

0       1       2       3       4       5       6       7       8   \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      51      52      53      54      55      56      57  \
0    0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1    0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2    0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3    0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4    0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   
..      ...  ...     ...     ...     ...     ...     ...     ...     ...   
203  0.2684  ...  0.0116  0.0098  0.0199  0.0033  0.0101  0.0065  0.0115   
204  0.2154  ...  0.0061  0.0093  0.0135  0.0063  0.0063  0.0034  0.0032   
205  0.2529  ...  0.0160  0.0029  0.0051  0.0062  0.0089  0.0140  0.0138   
206  0.2354  ...  0.0086  0.0046  0.0126  0.0036  0.0035  0.0034  0.0079   
207  0.2354  ...  0.0146  0.0129  0.0047  0.0039  0.0061  0.0040  0.0036   

         58      59  60  
0    0.0090  0.0032   R  
1    0.0052  0.0044   R  
2    0.0095  0.0078   R  
3    0.0040  0.0117   R  
4    0.0107  0.0094   R  
..      ...     ...  ..  
203  0.0193  0.0157   M  
204  0.0062  0.0067   M  
205  0.0077  0.0031   M  
206  0.0036  0.0048   M  
207  0.0061  0.0115   M  

[208 rows x 61 columns]

In [4]:
data.shape

(208, 61)

In [5]:
data.isnull().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [6]:
data.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57, 58, 59, 60],
           dtype='int64')

In [7]:
data[60].value_counts()

M    111
R     97
Name: 60, dtype: int64

In [8]:
X=data.drop(60,axis=1)
y=data[60]

In [9]:
X

0       1       2       3       4       5       6       7       8   \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      50      51      52      53      54      55      56  \
0    0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1    0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2    0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3    0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4    0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   
..      ...  ...     ...     ...     ...     ...     ...     ...     ...   
203  0.2684  ...  0.0203  0.0116  0.0098  0.0199  0.0033  0.0101  0.0065   
204  0.2154  ...  0.0051  0.0061  0.0093  0.0135  0.0063  0.0063  0.0034   
205  0.2529  ...  0.0155  0.0160  0.0029  0.0051  0.0062  0.0089  0.0140   
206  0.2354  ...  0.0042  0.0086  0.0046  0.0126  0.0036  0.0035  0.0034   
207  0.2354  ...  0.0181  0.0146  0.0129  0.0047  0.0039  0.0061  0.0040   

         57      58      59  
0    0.0084  0.0090  0.0032  
1    0.0049  0.0052  0.0044  
2    0.0164  0.0095  0.0078  
3    0.0044  0.0040  0.0117  
4    0.0048  0.0107  0.0094  
..      ...     ...     ...  
203  0.0115  0.0193  0.0157  
204  0.0032  0.0062  0.0067  
205  0.0138  0.0077  0.0031  
206  0.0079  0.0036  0.0048  
207  0.0036  0.0061  0.0115  

[208 rows x 60 columns]

In [10]:
y

0      R
1      R
2      R
3      R
4      R
      ..
203    M
204    M
205    M
206    M
207    M
Name: 60, Length: 208, dtype: object

In [11]:
y=pd.get_dummies(y,drop_first=True)

R--1

M--0

In [12]:
y

R
0    1
1    1
2    1
3    1
4    1
..  ..
203  0
204  0
205  0
206  0
207  0

[208 rows x 1 columns]

In [13]:
y.R.value_counts()

0    111
1     97
Name: R, dtype: int64

In [14]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=1)

In [15]:
X_train.shape

(166, 60)

In [16]:
X_test.shape

(42, 60)

Lets build a simple neural network model with two hidden layers

In [17]:
import tensorflow as tf
from tensorflow import keras

In [18]:
model=keras.Sequential([
    keras.layers.Dense(100,input_dim=60,activation='relu'),
    keras.layers.Dense(50,activation='relu'),
    keras.layers.Dense(1,activation='sigmoid')
])

In [19]:
model.compile(loss='BinaryCrossentropy',optimizer='adam',metrics=['accuracy'])

In [20]:
model.fit(X_train,y_train,epochs=10)

Epoch 1/10
6/6 [==============================] - 0s 997us/step - loss: 0.6636 - accuracy: 0.5964
Epoch 2/10
6/6 [==============================] - 0s 997us/step - loss: 0.6495 - accuracy: 0.6506
Epoch 3/10
6/6 [==============================] - 0s 1ms/step - loss: 0.6204 - accuracy: 0.6747
Epoch 4/10
6/6 [==============================] - 0s 997us/step - loss: 0.6190 - accuracy: 0.7530
Epoch 5/10
6/6 [==============================] - 0s 1ms/step - loss: 0.6014 - accuracy: 0.7470
Epoch 6/10
6/6 [==============================] - 0s 1ms/step - loss: 0.5784 - accuracy: 0.7289
Epoch 7/10
6/6 [==============================] - 0s 1ms/step - loss: 0.5660 - accuracy: 0.7229
Epoch 8/10
6/6 [==============================] - 0s 1ms/step - loss: 0.5499 - accuracy: 0.7470
Epoch 9/10
6/6 [==============================] - 0s 1ms/step - loss: 0.5476 - accuracy: 0.7952
Epoch 10/10
6/6 [==============================] - 0s 1ms/step - loss: 0.5184 - accuracy: 0.7289


In [21]:
model.evaluate(X_test,y_test)

2/2 [==============================] - 0s 998us/step - loss: 0.5359 - accuracy: 0.6905


[0.5358580350875854, 0.6904761791229248]

lets build a neural network model with three hidden layers

In [22]:
model1=keras.Sequential([
    keras.layers.Dense(100,input_dim=60,activation='relu'),
    keras.layers.Dense(50,activation='relu'),
    keras.layers.Dense(25,activation='relu'),
    keras.layers.Dense(1,activation='sigmoid')
])

In [23]:
model1.compile(loss='binary_crossentropy',optimizer='adam',metrics=['accuracy'])

In [24]:
model1.fit(X_train,y_train,epochs=50)

Epoch 1/50
6/6 [==============================] - 0s 1ms/step - loss: 0.6906 - accuracy: 0.4759
Epoch 2/50
6/6 [==============================] - 0s 1ms/step - loss: 0.6722 - accuracy: 0.5602
Epoch 3/50
6/6 [==============================] - 0s 1ms/step - loss: 0.6630 - accuracy: 0.5964
Epoch 4/50
6/6 [==============================] - 0s 1ms/step - loss: 0.6501 - accuracy: 0.5663
Epoch 5/50
6/6 [==============================] - 0s 1ms/step - loss: 0.6420 - accuracy: 0.5602
Epoch 6/50
6/6 [==============================] - 0s 1ms/step - loss: 0.6277 - accuracy: 0.6145
Epoch 7/50
6/6 [==============================] - 0s 1ms/step - loss: 0.6090 - accuracy: 0.6687
Epoch 8/50
6/6 [==============================] - 0s 1ms/step - loss: 0.5890 - accuracy: 0.6807
Epoch 9/50
6/6 [==============================] - 0s 1ms/step - loss: 0.5719 - accuracy: 0.6867
Epoch 10/50
6/6 [==============================] - 0s 2ms/step - loss: 0.5462 - accuracy: 0.7711
Epoch 11/50
6/6 [======================

With three hidden layers the model almost perfectly fits the data, lets test the model on the training data

In [25]:
model1.evaluate(X_test,y_test)

2/2 [==============================] - 0s 1ms/step - loss: 0.5864 - accuracy: 0.7143


[0.586426854133606, 0.7142857313156128]

It can be seen that the test accuracy is very low compared to the training accuracy

In [26]:
y_pred=model1.predict(X_test)

In [27]:
print(y_pred[:5])

[[6.8232577e-05]
 [8.0014765e-01]
 [9.6133316e-01]
 [3.1591952e-03]
 [9.9812877e-01]]


In [28]:
print(np.round(y_pred[:5]))

[[0.]
 [1.]
 [1.]
 [0.]
 [1.]]


In [29]:
print(y_test[:5])

     R
186  0
155  0
165  0
200  0
58   1


Testing accuracy of the model1 is leeser than the training accuracy

Lets Build a model with the dropout layer and see how the testing accuracy changes

In [30]:
model2=keras.Sequential([
    keras.layers.Dense(100,input_dim=60,activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(50,activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(25,activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1,activation='sigmoid')
])

In [31]:
model2.compile(loss='binary_crossentropy',optimizer='adam',metrics=['accuracy'])

In [32]:
model2.fit(X_train,y_train,epochs=50)

Epoch 1/50
6/6 [==============================] - 0s 1ms/step - loss: 0.7752 - accuracy: 0.5000
Epoch 2/50
6/6 [==============================] - 0s 1ms/step - loss: 0.7276 - accuracy: 0.4819
Epoch 3/50
6/6 [==============================] - 0s 1ms/step - loss: 0.7351 - accuracy: 0.5000
Epoch 4/50
6/6 [==============================] - 0s 1ms/step - loss: 0.7126 - accuracy: 0.4759
Epoch 5/50
6/6 [==============================] - 0s 2ms/step - loss: 0.7024 - accuracy: 0.5422
Epoch 6/50
6/6 [==============================] - 0s 1ms/step - loss: 0.6710 - accuracy: 0.5964
Epoch 7/50
6/6 [==============================] - 0s 2ms/step - loss: 0.6807 - accuracy: 0.5663
Epoch 8/50
6/6 [==============================] - 0s 2ms/step - loss: 0.6843 - accuracy: 0.5602
Epoch 9/50
6/6 [==============================] - 0s 1ms/step - loss: 0.7084 - accuracy: 0.5120
Epoch 10/50
6/6 [==============================] - 0s 2ms/step - loss: 0.6791 - accuracy: 0.5482
Epoch 11/50
6/6 [======================

In [33]:
model2.evaluate(X_test,y_test)

2/2 [==============================] - 0s 997us/step - loss: 0.4860 - accuracy: 0.7857


[0.4860021471977234, 0.7857142686843872]

testing accuracy increased from 71% to 78% after introducing the dropout layers